lib/experimental/register-webfonts-from-theme-json.php

```diff
@@ -42,133 +41,48 @@ function gutenberg_register_webfonts_from_theme_json() {
 		return;
 	}
 
-	$webfonts = array();
-
-	// Look for fontFamilies.
-	foreach ( $settings['typography']['fontFamilies'] as $font_families ) {
-		foreach ( $font_families as $font_family ) {
-
-			// Skip if fontFace is not defined.
-			if ( empty( $font_family['fontFace'] ) ) {
-				continue;
-			}
+	$font_faces_to_register = array();
 
-			$font_family['fontFace'] = (array) $font_family['fontFace'];
+	foreach ( $settings['typography']['fontFamilies'] as $font_families_by_origin ) {
+		foreach ( $font_families_by_origin as $font_family ) {
+			if ( isset( $font_family['provider'] ) ) {
+				if ( empty( $font_family['fontFaces'] ) ) {
+					trigger_error(
+						sprintf(
+							/* translators: %s: font family name */
+							esc_html__( 'The "%s" font family specifies a provider, but no font faces.', 'gutenberg' ),
+							wp_kses_data( $font_family['fontFamily'] )
+						)
+					);
 
-			foreach ( $font_family['fontFace'] as $font_face ) {
-				// Skip if the webfont was registered through the Webfonts API.
-				if ( isset( $font_face['origin'] ) && 'gutenberg_wp_webfonts_api' === $font_face['origin'] ) {
 					continue;
 				}
 
-				// Check if webfonts have a "src" param, and if they do account for the use of "file:./".
-				if ( ! empty( $font_face['src'] ) ) {
-					$font_face['src'] = (array) $font_face['src'];
-
-					foreach ( $font_face['src'] as $src_key => $url ) {
-						// Tweak the URL to be relative to the theme root.
-						if ( ! str_starts_with( $url, 'file:./' ) ) {
-							continue;
-						}
-						$font_face['src'][ $src_key ] = get_theme_file_uri( str_replace( 'file:./', '', $url ) );
-					}
-				}
+				foreach ( $font_family['fontFaces'] as $font_face ) {
+					$font_face['provider'] = $font_family['provider'];
+					$font_face             = gutenberg_resolve_font_face_uri( $font_face );
+					$font_face             = gutenberg_webfont_to_kebab_case( $font_face );
 
-				// Convert keys to kebab-case.
-				foreach ( $font_face as $property => $value ) {
-					$kebab_case               = _wp_to_kebab_case( $property );
-					$font_face[ $kebab_case ] = $value;
-					if ( $kebab_case !== $property ) {
-						unset( $font_face[ $property ] );
-					}
+					$font_faces_to_register[] = $font_face;
 				}
 
-				$webfonts[] = $font_face;
+				continue;
 			}
-		}
-	}
-	foreach ( $webfonts as $webfont ) {
-		wp_webfonts()->register_webfont( $webfont );
-		wp_webfonts()->enqueue_webfont( $webfont['font-family'] );
```

Comment:
@aristath now that we're allow-listing through entries in `theme.json`, it doesn't make sense to enqueue a webfont from the get-go. Enqueueing happens elsewhere (`enqueue-webfonts-listed-in-theme-json`). This is for separation of concerns. Let me know if it makes you unhappy.

Reply:
That's OK, provided we get the user data as well and not just the theme data as noted on https://github.com/WordPress/gutenberg/pull/39988#discussion_r848263020, so that variations keep working 👍